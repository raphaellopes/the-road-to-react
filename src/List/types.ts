import { StoriesType, StoryType } from '../types';

export type ListProps = {
  list: StoriesType;
  onRemoveItem: (item:StoryType) => void;
}

export type ItemProps = {
  item: StoryType,
  onRemoveItem: (item:StoryType) => void;
};

export type SortTypes = 'TITLE' | 'AUTHOR' | 'COMMENT' | 'POINT' | 'NONE';

export type SortDictType = {
  [key:string]: (data:StoriesType) => StoriesType
};
