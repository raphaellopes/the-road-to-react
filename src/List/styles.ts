import styled from 'styled-components';

export const ItemColumn = styled.span`
  ${({ width }: { width: string }) => `
    width: ${width};
    padding: 0 5px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;

    > a {
      color: inherit;
    }
  `}
`

export const ItemContainer = styled.div`
  display: flex;
  align-items: center;
  padding-bottom: 5px;
`;
