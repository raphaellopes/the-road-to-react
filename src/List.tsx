import React from 'react';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { StoriesType, StoryType } from './types';
import { ButtonSmall } from './styles';

type ListProps = {
  list: StoriesType;
  onRemoveItem: (item:StoryType) => void;
}

const List = ({ list, onRemoveItem }:ListProps) => (
  <>
    {list.map(
      (item) => (
        <Item key={item.objectID} onRemoveItem={onRemoveItem} item={item} />
      )
    )}
  </>
);

const ItemColumn = styled.span`
  ${({ width }: { width: string }) => `
    width: ${width};
    padding: 0 5px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;

    > a {
      color: inherit;
    }
  `}
`

const ItemStyled = styled.div`
  display: flex;
  align-items: center;
  padding-bottom: 5px;
`;
type ItemProps = {
  item: StoryType,
  onRemoveItem: (item:StoryType) => void;
};
export const Item = ({
  item,
  onRemoveItem
}: ItemProps) => {
  const { url, title, author, num_comments, points } = item;
  return (
    <ItemStyled>
      <ItemColumn width="40%">
        <a href={url}>{title}</a>
      </ItemColumn>
      <ItemColumn width="30%">{author}</ItemColumn>
      <ItemColumn width="10%">{num_comments}</ItemColumn>
      <ItemColumn width="10%">{points}</ItemColumn>
      <ItemColumn width="10%">
        <ButtonSmall
          type="button"
          onClick={() => onRemoveItem(item)}>
          <FontAwesomeIcon icon="check" data-testid="dismiss" />
        </ButtonSmall>
      </ItemColumn>
    </ItemStyled>
  );
}

export default List;
