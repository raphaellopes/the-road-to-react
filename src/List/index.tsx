import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { ButtonSmall } from '../styles';
import { ListProps, ItemProps } from './types';
import { ItemContainer, ItemColumn } from './styles';

const List = ({ list, onRemoveItem }:ListProps) => (
  <>
    {list.map(
      (item) => (
        <Item key={item.objectID} onRemoveItem={onRemoveItem} item={item} />
      )
    )}
  </>
);

export const Item = ({
  item,
  onRemoveItem
}: ItemProps) => {
  const { url, title, author, num_comments, points } = item;
  return (
    <ItemContainer>
      <ItemColumn width="40%">
        <a href={url}>{title}</a>
      </ItemColumn>
      <ItemColumn width="30%">{author}</ItemColumn>
      <ItemColumn width="10%">{num_comments}</ItemColumn>
      <ItemColumn width="10%">{points}</ItemColumn>
      <ItemColumn width="10%">
        <ButtonSmall
          type="button"
          onClick={() => onRemoveItem(item)}>
          <FontAwesomeIcon icon="check" data-testid="dismiss" />
        </ButtonSmall>
      </ItemColumn>
    </ItemContainer>
  );
}

export default List;
