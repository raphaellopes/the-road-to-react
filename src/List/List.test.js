import { render, screen, fireEvent } from '@testing-library/react';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faCheck, faArrowUp, faArrowDown } from '@fortawesome/free-solid-svg-icons';
import { stories, storyOne } from '../mockedTests';

import List, { Item } from './index';

library.add(faCheck, faArrowUp, faArrowDown);

describe('List', () => {
  const listProps = {
    list: stories,
    onRemoveItem: jest.fn(),
  };

  test('renders snapshot', () => {
    const { container } = render(<List {...listProps} />)
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders the list with its properties', () => {
    render(<List {...listProps} />);
    expect(screen.getAllByTestId('dismiss').length).toBe(2);
  });

  test('calls onRemoveItem for first item', () => {
    render(<List {...listProps} />);
    fireEvent.click(screen.getAllByTestId('dismiss')[0]);
    expect(listProps.onRemoveItem).toHaveBeenCalledTimes(1);
  });
});

describe('Item', () => {
  test('renders all properties', () => {
    render(<Item item={storyOne} />);
    expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
    expect(screen.getByText('React')).toHaveAttribute(
      'href',
      'https://reactjs.org/'
    );
  });

  test('renders a clickable dismiss button', () => {
    render(<Item item={storyOne} />);
    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  test('clicking the dismiss button calls the callback handler', () => {
    const handleRemoveItem = jest.fn();
    render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);
    fireEvent.click(screen.getByRole('button'));
    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
  });

  test('renders snapshot', () => {
    const { container } = render(<Item item={storyOne} />)
    expect(container.firstChild).toMatchSnapshot();
  });
});
