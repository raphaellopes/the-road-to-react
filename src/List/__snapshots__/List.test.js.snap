// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
.c4 {
  background: transparent;
  border: 1px solid #171212;
  padding: 5px;
  cursor: pointer;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
}

.c4:hover {
  background: #171212;
  color: #ffffff;
}

.c4:hover > svg > g {
  fill: #ffffff;
  stroke: #ffffff;
}

.c5 {
  padding: 5px;
}

.c1 {
  width: 40%;
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 > a {
  color: inherit;
}

.c2 {
  width: 30%;
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 > a {
  color: inherit;
}

.c3 {
  width: 10%;
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c3 > a {
  color: inherit;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-bottom: 5px;
}

<div
  class="c0"
>
  <span
    class="c1"
    width="40%"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    class="c2"
    width="30%"
  >
    Jordan Walke
  </span>
  <span
    class="c3"
    width="10%"
  >
    3
  </span>
  <span
    class="c3"
    width="10%"
  >
    4
  </span>
  <span
    class="c3"
    width="10%"
  >
    <button
      class="c4 c5"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check fa-w-16 "
        data-icon="check"
        data-prefix="fas"
        data-testid="dismiss"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </button>
  </span>
</div>
`;

exports[`List renders snapshot 1`] = `
.c4 {
  background: transparent;
  border: 1px solid #171212;
  padding: 5px;
  cursor: pointer;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
}

.c4:hover {
  background: #171212;
  color: #ffffff;
}

.c4:hover > svg > g {
  fill: #ffffff;
  stroke: #ffffff;
}

.c5 {
  padding: 5px;
}

.c1 {
  width: 40%;
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 > a {
  color: inherit;
}

.c2 {
  width: 30%;
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 > a {
  color: inherit;
}

.c3 {
  width: 10%;
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c3 > a {
  color: inherit;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-bottom: 5px;
}

<div
  class="c0"
>
  <span
    class="c1"
    width="40%"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    class="c2"
    width="30%"
  >
    Jordan Walke
  </span>
  <span
    class="c3"
    width="10%"
  >
    3
  </span>
  <span
    class="c3"
    width="10%"
  >
    4
  </span>
  <span
    class="c3"
    width="10%"
  >
    <button
      class="c4 c5"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check fa-w-16 "
        data-icon="check"
        data-prefix="fas"
        data-testid="dismiss"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
    </button>
  </span>
</div>
`;
